ENH: show warning when np.maximum receives more than 2 inputs

lvllvl opened the pull request:
attempts to resolve #27639
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

jorenham commented on numpy/_typing/_ufunc.pyi:
`out` can be passed positionally by keyword in `ufunc`s, so these changes should be reverted

seberg reviewed (changes requested):
I wouldn't mind discussing whether we should start forcing `out=` more generally, since the overhead of using `out=` is pretty low now.

But maybe we say maximum is special, which is fine:
* You are doing twice the same thing in the same code.
* Do **not** just go to an error, this should go through a `DeprecationWarning`.  Maybe one can argue that it is almost always a bug and we go directly to a `VisibleDeprecationWarning`.  But I am a bit skeptical, also because `DeprecationWarnings` are seen more these days anyway.

mhvk reviewed (commented):
I don't think this should happen just for `maximum` (though that discussion is perhaps best done in #27639).

But it certainly shouldn't be in the hot path for every ufunc -- instead, on l.4321 (original one), one could do
```
if NPY_UNLIKELY(len_args != nin) {
    if (len_args <= nop) {
       /* new deprecation stuff */
    }
    else {
       PyErrFormat(...);
       goto fail;
    } 
}
```

seberg reviewed (changes requested):
Thanks, but this needs some work, also you didn't address the discussion around minimum and maximum.

lvllvl commented on numpy/_core/src/umath/ufunc_object.c:
```diff
@@ -4325,7 +4343,28 @@ ufunc_generic_fastcall(PyUFuncObject *ufunc,
         goto fail;
     }
+    /* Extra positional args but *no* keywords? */
+    if (len_args > nin && (kwnames == NULL || PyTuple_GET_SIZE(kwnames) == 0)) {
```
Ok removing this

lvllvl commented on numpy/_core/src/umath/ufunc_object.c:
```diff
@@ -4325,7 +4343,28 @@ ufunc_generic_fastcall(PyUFuncObject *ufunc,
+
+        /* legal out? -> skip */
+        if (PyArray_Check(extra) || extra == Py_None ||
+            _is_sequence_of_arrays_or_None(extra)) {
```
removing this logic

lvllvl commented:
> Thanks, but this needs some work, also you didn't address the discussion around minimum and maximum.

Hi @seberg, 

I added some changes related to the discussion around maximum/minimum, e.g., 
- when 2 or more positional args are pass to `np.maximum`, `np.minimum` there will be a `DeprecationWarning`
- the warning will be about the positional `out`, the behavior will otherwise be unchanged

Let me know if that addresses all your concerns. I can add other changes as well. I don't know if changing documentation would be in this scope but I can add that as well, if needed.

seberg commented on numpy/_core/src/umath/umathmodule.c:
```diff
@@ -279,6 +281,18 @@ int initumath(PyObject *m)
+    if (PyDict_GetItemStringRef(d, "maximum", &tmp) < 0) {
+        return -1;
+    }
+    UFUNC_MAXIMUM = (PyUFuncObject *)tmp;
```
Just store it directly?

seberg replied:
The old comment still applies again now: Just store it directly into UFUNC_MAXIMUM.

But I am not actually sure this is perfectly thread-safe. Can you try getting these from the n_ops struct in number.h. (I think we should use it, if including it this way around is awkward, it might actually make sense to move it into umathmodule.c)

lvllvl replied:
ok I'll try to use the n_ops from `number.h` first. It seems to work, I just need to cast ufunc, e.g., 
`        if ((PyObject *)ufunc == n_ops.maximum || (PyObject *)ufunc == n_ops.minimum) {`

seberg commented on numpy/_core/tests/test_umath.py:
```diff
@@ -2443,6 +2444,23 @@ def test_precision(self):
+        a = np.array([1])
+        b = np.array([2])
+        out = np.empty_like(a)
+        with pytest.warns(DeprecationWarning, match=r"Passing more than 2 positional arguments to np\.maximum is deprecated"):
```
This is failing the linter for good reason.

seberg commented on numpy/_core/src/umath/ufunc_object.c:
```diff
@@ -4363,6 +4363,20 @@ ufunc_generic_fastcall(PyUFuncObject *ufunc,
         }
+
+        /* Extra positional args but *no* keywords */
+        if ((ufunc == UFUNC_MAXIMUM || ufunc == UFUNC_MINIMUM) && !all_none) {
```
This is a bit tautological, you already have both `ufunc ==` checks just below.  TBH, I think it would be fine to just make one path though and don't care about the duplication.  (or use `ufunc->name`.)

seberg commented on numpy/_core/tests/test_umath.py:
```diff
@@ -2443,6 +2444,23 @@ def test_precision(self):
+        out = np.empty_like(a)
+
+        with pytest.warns(DeprecationWarning, match=r"Passing more than 2 positional arguments to np\.minimum is deprecated"):
+            result = np.minimum(a, b, out)
```
Sorry for not flagging earlier, but please move this test (following the pattern there) to `test_deprecations.py`.

In general I don't care too much about it, but I do care about testing the error path when the warning is raised as an error, and that is not tested here, but the pattern in the file always tests it.

(And when you do that, you should see that the C code is missing error checks and the test fails in weird ways.)

seberg reviewed (commented):
I suppose @mhvk was also mildly in favor of just doing this (limited to minimum/maximum), so this looks pretty close, thanks.

lvllvl committed:
ENH: updates based on feedback

lvllvl committed:
ENH: more changes based on feedback

lvllvl committed:
ENH: updating the error checks

mhvk reviewed (commented):
Agreed that this is mostly done - just small ideas for improvements.

lvllvl commented on numpy/_core/src/umath/ufunc_object.c:
```diff
@@ -4368,6 +4368,22 @@ ufunc_generic_fastcall(PyUFuncObject *ufunc,
+
+        /* Extra positional args but *no* keywords */
+        /* DEPRECATED NumPy 2.4, 2025-08 */
+        if (strcmp(ufunc->name, "maximum") == 0 || strcmp(ufunc->name, "minimum") == 0) {
```
Ok, my mistake I should've clarified before changing back to `strcmp`. I can change it back if you'd like.

I simplified the `if-conditional` like @mhvk suggested and I think I'll stick with the `DEPRECATE` macro, if you don't mind.

lvllvl committed:
ENH: simplify tests, remove snprintf/PyWarn_Format

lvllvl committed:
ENH: re-insert faster comparison

lvllvl committed:
ENH: use n_ops struct for comparison

lvllvl committed:
DEP: resolve conflicts

seberg reviewed (commented):
Sorry, it seems we did forget one thing: A deprecation should probably have a very brief release note.

Everything else looks fine though, just prefer less empty lines :).

lvllvl commented on numpy/_core/src/umath/ufunc_object.c:
```diff
@@ -4368,6 +4369,21 @@ ufunc_generic_fastcall(PyUFuncObject *ufunc,
+        /* Extra positional args but no keywords */
+        /* DEPRECATED NumPy 2.4, 2025-08 */
+        if ((PyObject *)ufunc == n_ops.maximum || (PyObject *)ufunc == n_ops.minimum) {
+            
```
I don't see any changes here. Do you mind writing this out again? I'll try to access this PR a different way, maybe your suggestion will appear that way.

lvllvl committed:
DEP: adding release note / updates based on feedback

lvllvl committed:
DEP: formatting changes on release note

lvllvl committed:
DEP: formatting changes on release note2
